src/Relation/Nullary/Finite/Setoid.agda

```diff
@@ -0,0 +1,148 @@
+open import Data.Product.Base using (_×_; _,_; swap; Σ-syntax; proj₁; ∃)
+open import Data.Unit.Base using (⊤)
+open import Function.Base using (id; _∘_)
+open import Function.Bundles using (Func; Injection; Surjection; Inverse)
+open import Level using (Level; _⊔_; 0ℓ) renaming (suc to lsuc)
+open import Relation.Binary.Core using (Rel)
+open import Relation.Binary.Bundles using (Setoid)
+open import Relation.Binary.Structures using (IsEquivalence)
+import Relation.Binary.Construct.On as On
+import Relation.Binary.PropositionalEquality.Properties as ≡
+
+private
+  variable
+    c ℓ c′ ℓ′ : Level
+
+Ω : ∀ p → Setoid (lsuc p) p
+Ω p = record
+  { Carrier = Set p
+  ; _≈_ = λ P Q → (P → Q) × (Q → P)
+  ; isEquivalence = record
+    { refl = id , id
+    ; sym = swap
+    ; trans = λ (f , g) (f′ , g′) → f′ ∘ f , g ∘ g′
+    }
+  }
+
+Subset : Setoid c ℓ → (p : Level) → Set (c ⊔ ℓ ⊔ lsuc p)
+Subset X p = Func X (Ω p)
+
+FullSubset : {X : Setoid c ℓ} → Subset X 0ℓ
+FullSubset = record { to = λ _ → ⊤ }
+
+record EquivalenceRelation (X : Setoid c ℓ) r : Set (c ⊔ ℓ ⊔ lsuc r) where
+  infix 4 _∼_
+  open Setoid X
+  field
+    _∼_ : Rel Carrier r
+    ∼-resp-≈ : ∀ {x x′ y y′} → x ≈ x′ → y ≈ y′ → x ∼ y → x′ ∼ y′
+    ∼-isEquivalence : IsEquivalence _∼_
+
+  open IsEquivalence ∼-isEquivalence public renaming
+    ( refl                 to ∼-refl
+    ; trans                to ∼-trans
+    ; sym                  to ∼-sym
+    ; reflexive            to ∼-reflexive
+    ; isPartialEquivalence to ∼-isPartialEquivalence
+    )
+
+  ≈⇒∼ : ∀ {x y} → x ≈ y → x ∼ y
+  ≈⇒∼ q = ∼-resp-≈ refl q ∼-refl
+
+_⁻¹[_] : ∀ {p} {X : Setoid c ℓ} {Y : Setoid c′ ℓ′} →
+  Func X Y → Subset Y p → Setoid (c ⊔ p) ℓ
+_⁻¹[_] {X = X} f S =
+  On.setoid {B = Σ[ x ∈ X.Carrier ] S.to (f.to x)} X proj₁
+  where
+  module X = Setoid X
+  module f = Func f
+  module S = Func S
+
+_[_] : ∀ {p} {X : Setoid c ℓ} {Y : Setoid c′ ℓ′} →
+  Func X Y → Subset X p → Setoid (c ⊔ p) ℓ′
+_[_] {X = X} {Y} f S =
+  On.setoid {B = Σ[ x ∈ X.Carrier ] S.to x} Y (f.to ∘ proj₁)
+  where
+  module X = Setoid X
+  module f = Func f
+  module S = Func S
+
+↣⇒⊆ : ∀ {X : Setoid c ℓ} {Y : Setoid c′ ℓ′} → Injection Y X → Subset X (c′ ⊔ ℓ)
+↣⇒⊆ {X = X} {Y} m = record
+  { to = λ x → ∃ \ y → to y X.≈ x
+  ; cong = λ p →
+    (λ (y , q) → y , X.trans q p) , (λ (y , q) → y , X.trans q (X.sym p))
+  }
+  where
+  module X = Setoid X
+  module Y = Setoid Y
+  open Injection m
+
+_/_ : ∀ {r} (X : Setoid c ℓ) → EquivalenceRelation X r → Setoid c r
+X / R = record
+  { Carrier = Carrier
+  ; _≈_ = _∼_
+  ; isEquivalence = ∼-isEquivalence
+  }
+  where
+  open Setoid X
+  open EquivalenceRelation R
+
+include-/ : ∀ {r} {X : Setoid c ℓ} (R : EquivalenceRelation X r) →
+  Surjection X (X / R)
+include-/ {X = X} R = record
+  { to = id
+  ; cong = ≈⇒∼
+  ; surjective = λ y → y , ∼-refl
+  }
+  where
+  open Setoid X
+  open EquivalenceRelation R
+
+record IsStronglyFinite (X : Setoid c ℓ) (n : ℕ) : Set (c ⊔ ℓ) where
```

Comment:
Above these, there should be comments as to the origin of the names (with links).

Reply:
I could not find an origin for the name `StronglyFinite`. There are a few references that define finite sets or a notion of finiteness but I could not see any of them using `StronglyFinite` in particular -

- [Dependently typed programming with finite sets ](https://dl.acm.org/citation.cfm?id=2808102)
- [Brent's Yorgey's PhD thesis](https://github.com/byorgey/thesis)
- [Towards Certifiable Implementation of Graph Transformation via Relation Categories](https://link.springer.com/chapter/10.1007/978-3-642-33314-9_6)

Which one should I cite? Or should it be the GitHub comment where `StronglyFinite` popped up for the first time?

Reply:
Probably `FinitelyEnumerable` is closer to what's used out there.

Reply:
Loking at the references, [Dependently typed programming with finite sets](https://dl.acm.org/citation.cfm?id=2808102) introduces "finitely enumerable" for the first time. It classifies the notion of finiteness using some other references -

> Intuitionistic frameworks give rise to a rich variety of notions of finiteness that collapse classically. In [[8]](https://inria.hal.science/inria-00503917/file/Coquand_Spiwack_-_2010_-_Constructively_Finite.pdf), [[5]](https://arxiv.org/pdf/1210.1200.pdf) and [[14]](https://www.irif.fr/~letouzey/types2014/slides-39.pdf), the author describe various concepts of finiteness and their interrrelation. According to their classification, this paper focuses on the strongest notion of finiteness, namely, finitely enumerable (listable) sets.

[[8]](https://inria.hal.science/inria-00503917/file/Coquand_Spiwack_-_2010_-_Constructively_Finite.pdf) talks about "Enumerated" sets which I found to be the closest here.

> A set A will be said to be enumerated, or simply finite, noted A ε F when there is a list of all its elements. We call such a list an enumeration of A.

I also found https://github.com/agda/agda-finite-prover, which looks somewhat relevant here.

Reply:
`Enumerated` is another excellent name.

Reply:
Spiwack and Coqand unfortunately use rather vague wording in our definition of interest:

> **Definition**
> A set *A* will be said to be *enumerated*, or simply finite, noted *A* *ε* 𝓕 when there
> is a list of all its elements. We call such a list an *enumeration* of *A*.

I *think* this means ∃*xs* : List *A*. ∀*x* : *A*. *x* ∈ *xs*, which allows for duplicates in the list. This is equivalent to `FinitelyEnumerable` from this PR:

**Proof**
(enumerated → `FinitelyEnumerable`): Let `size` be the length of the list *xs*. The function is given by `lookup`. This is surjective by the fact that for every *x*, we have an *e* : *x* ∈ *xs*, and we can take the index of *e*, which `lookup` maps to *x*.
(`FinitelyEnumerable` → enumerated): The list is given by the tabulation of the surjective function. Given an *x* : *A*, it is in the tabulation by surjectivity.

I think this discussion, if anything, motivates the name `StronglyFinite`. It is stronger than all of these other commonly studied notions of finiteness, and is perhaps the strongest notion we could reasonably call “finiteness”. Definitely, if we had univalence, then any property stronger than “∃ `n`. it's equivalent to `Fin n`” precludes the type `Fin n` for some `n`. If `StronglyFinite` is strictly stronger than the other notions of finiteness being discussed, we could ask whether it's worth having in the library at all. I think it is. For example, when people talk about categories whose objects are finite sets, I think they nearly always mean strongly finite sets. Also, I can imagine that strong finiteness is too trivial to mention in pen-and-paper mathematics, because you could just say Fin *n* instead, but I think people genuinely do want to talk about sets which are strongly finite in Agda.

Anyway, it's important that we get these things right, so thanks for doing the research. I don't have a good enough intuition on the three later (weaker) definitions from the same paper, but hopefully one of them is equivalent to being subfinitely enumerable (if I had to put money on it, I'd go with bounded size, because it seems to have similar properties). I guess it would be good to have all of these list-based definitions in the library too, connected with the appropriate equivalence/inclusion proofs to the definitions already in this PR. It would help get rid of some definitions without proofs.

Reply:
I certainly very happy to stick to `StronglyFinite` !